Rewrite metadata store

Random-Liu opened the pull request:
Previously, containerd only managed running container, so we have to do checkpoint to maintain complete lifecycle of a container and sandbox. That's why we introduced the metadata store, which was targeted to be a checkpoint store for container, sandbox and image if needed.

However, it seems that we should not rely on the metadata store now, because:
1) Containerd added container metadata store, and also supports container label, which could help us checkpoint constant data. We only need to maintain checkpoint of container status now.
2) The metadata store was originally designed for pure data checkpoint, it only supports update via `Update` function, which has some limitations:
  a) `Update` function is not flexible enough, e.g. for image, we need `CreateOrUpdate` operation to create a metadata if it does not exsit, or update it if does.
  b) `Update` and checkpoint doesn't make sense to some in-memory data we need to associate with the metadata lifecycle, e.g. containerd container/image client, cni netns object, stop container channel etc.
  c) The whole metadata store logic is unnecessarily complex given that we don't need a general checkpoint solution now.

Thus I made this change. Relatively large, but help us avoid future pain.

Use container store as an example, the store is organized like this:
```
Store (a simple in-memory map)
   |--> Container (a collection of all resources associated with the container)
             | --> Metadata (constant metadata of the container, will be checkpointed in containerd container label)
             | --> Status (container status maintained in our own checkpoint, the only checkpoint we need now)
             | --> Containerd Container Client (container client of containerd, added soon)
             | --> Container StopCh (a channel indicates container is stopped or not, added soon)
             | --> More... (If needed in the future)
```

For sandbox, we don't need status.
For image, we don't event need a separate `Metadata` field, because containerd provided everything we need.

Random-Liu commented:
CRI Validation test result is not changed with this PR:
```
Ran 36 of 36 Specs in 94.648 seconds
FAIL! -- 23 Passed | 13 Failed | 0 Pending | 0 Skipped --- FAIL: TestE2ECRI (94.65s)
FAIL
```

mikebrow commented on hack/verify-lint.sh:
```diff
@@ -24,6 +24,7 @@ for d in $(find . -type d -a \( -iwholename './pkg*' -o -iwholename './cmd*' \))
 		 --exclude='.*_test\.go:.*error return value not checked.*\(errcheck\)$' \
 		 --exclude='duplicate of.*_test.go.*\(dupl\)$' \
 		 --exclude='.*/mock_.*\.go:.*\(golint\)$' \
+		 --exclude='declaration of "err" shadows declaration at.*\(vetshadow\)' \
```
Removing will introduce the possibility of programming errors with `err` vars.

Random-Liu replied:
The problem is that the check is known as error prone. It sometimes report error and sometimes not, and I don't think something like:
```
if err := func1(); err != nil {
}
if err := func2(); err != nil {
}
```
should be an error.

mikebrow replied:
only with err? Or are you arguing for a different recommended pattern of use of variables in go?

Random-Liu replied:
Only with err. I mean.

mikebrow replied:
Obviously `err := foo` is just a shortcut for `var err error =` In your above use case you are repetitivly stating var err error, when all you are really doing is assigning with an =. In your use case the cleaner way is to declare first then just use err = since you don't like using := sometimes and = other times, yes?

mikebrow replied:
I guess my question is then, why is err a special use case of variable declaration and assignment? I understand the thought for keeping each test independent but the variable redeclaration doesn't make the test more independent than using assignment. Symmetry?

Random-Liu replied:
Offline discussed. Keep the check and update the code.

mikebrow commented:
When you refactor you really refactor! :-) Will review from the perspective of it being a complete replacement.

yujuhong commented:
Will review today.

Random-Liu commented:
@yujuhong Thanks~

yujuhong commented on pkg/store/container/status.go:
```diff
@@ -0,0 +1,146 @@
+}
+
+// Delete deletes the container status from disk atomically.
+func (m *statusStorage) Delete() error {
```
Why can't we delete `m.status` in this function? It looks like `Delete()` is only meaningful if we back the data on the disk.

Random-Liu replied:
It doesn't matter whether we delete `m.status` or not, because once the container is removed from container list, no one should be able to retrieve and access it.

Delete `m.status` will only cause race condition between operations which already get the reference.

yujuhong commented on pkg/store/image/image.go:
```diff
@@ -0,0 +1,120 @@
+
+// NewStore creates an image store.
+func NewStore() *Store {
+	return &Store{images: make(map[string]Image)}
```
Same question about the rationale of storing `Image` vs `*Image`

Random-Liu replied:
Same with above.

yujuhong commented on pkg/store/image/image.go:
```diff
@@ -0,0 +1,120 @@
+)
+
+// Image contains all resources associated with the image. All fields
+// are **read-only**.
```
Is that true? In `Add()`, you merge RepoTags and RepoDigest and write it back to the store.

Random-Liu replied:
Will update the comment to "All fields **MUST** not be mutated directly".

mikebrow replied:
Can we do this change universally on the other read only comments?

Random-Liu replied:
Done.

yujuhong commented on pkg/store/container/container.go:
```diff
@@ -0,0 +1,108 @@
+)
+
+// Container contains all resources associated with the container. All fields of
+// Container are thread safe.
```
I'm not sure how fields can be thread-safe. I think you mean the methods to mutate the status are thread-safe?

Random-Liu replied:
Done.

yujuhong commented on pkg/store/sandbox/sandbox.go:
```diff
@@ -0,0 +1,88 @@
+// Store stores all sandboxes.
+type Store struct {
+	lock      sync.RWMutex
+	sandboxes map[string]Sandbox
```
Same question about storing pointers vs. structs.

Random-Liu replied:
Same.

Random-Liu commented:
Will rebase this and address comments.

Random-Liu commented:
Moved to v0.2.0, because this PR mainly:
1) Eases the work to switching to new containerd client, and add permanent network namespace.
2) Make it possible to checkpoint most metadata in container label.
3) Make it easier to make future change.

There is no new feature introduced.

Since we don't have time to finish the containerd client switch and permanent network namespace work. Let's also punt this one to v0.2.0.

Random-Liu commented on pkg/store/container/container.go:
```diff
@@ -0,0 +1,108 @@
+
+// NewStore creates a container store.
+func NewStore() *Store {
+	return &Store{containers: make(map[string]Container)}
```
I use `Container` rather than `*Container` because:
1) The value copy could prevent callers from mutating the struct fields in some way, because most fields are primitive types.
2) Most importantly, when caller using the api, usually they won't expect to be able to mutate fields of a returned value, but may expect to mutate a returned pointer.

As for the overhead, metadata is small, so there shouldn't be too much overhead. :)

Random-Liu commented:
@yujuhong Rebased the PR. Could you help me take another look? Maybe focus on the first commit.

@mikebrow Could you help me review this PR?

mikebrow commented on pkg/server/container_create.go:
```diff
@@ -158,12 +158,23 @@ func (c *criContainerdService) CreateContainer(ctx context.Context, r *runtime.C
-		return nil, fmt.Errorf("failed to add container metadata %+v into store: %v",
-			meta, err)
+	if err := c.containerStore.Add(container); err != nil {
+		return nil, fmt.Errorf("failed to add container %q into store: %v", id, err)
```
I see the storing of the status in the NewContainer() call but don't understand why the status is stored before the container. Is that to make sure the container always has status? maybe the container store.add could be changed to add container with status to make it more atomic?

Random-Liu replied:
`Add` need to lock the container list. I don't think we want to lock the list and do some disk operation.

Random-Liu replied:
Done. Added `Container.Delete` function.

mikebrow commented on pkg/store/container/container.go:
```diff
@@ -0,0 +1,108 @@
+	Metadata
+	// Status stores the status of the container.
+	Status StatusStorage
+	// TODO(random-liu): Add containerd container client.
```
?

Random-Liu replied:
I mean in the future, we could [this](https://github.com/containerd/containerd/blob/master/container.go#L22) here, after we start to use containerd client.

mikebrow replied:
kk

mikebrow commented on pkg/store/container/metadata.go:
```diff
@@ -0,0 +1,76 @@
+)
+
+// NOTE(random-liu):
+// 1) Metadata is read-only.
```
? you mean can't be updated after create?

Random-Liu replied:
Yeah, will update.

Random-Liu replied:
Done.

mikebrow commented on pkg/store/container/metadata.go:
```diff
@@ -0,0 +1,76 @@
+
+// NOTE(random-liu):
+// 1) Metadata is read-only.
+// 2) Metadata is checkpointed as containerd container label.
```
just confirming we'll use labels not annotations for all metadata so we can find the container by the metadata labels or... maybe some combination of labels and annotations...?

Random-Liu replied:
containerd only has label, there is no notion of annotation in containerd.

mikebrow replied:
oci will have the annotations.. and I suspect containerd will let them pass through without providing first class support

Random-Liu commented on pkg/metadata/container.go:
```diff
@@ -1,183 +0,0 @@
-/*
```
You won't see the diff even after I merge those 2. Will change to "rewrite".

Random-Liu commented on pkg/server/container_execsync.go:
```diff
@@ -45,15 +45,16 @@ func (c *criContainerdService) ExecSync(ctx context.Context, r *runtime.ExecSync
 
-	// Get container metadata from our container store.
-	meta, err := c.containerStore.Get(r.GetContainerId())
+	// Get container from our container store.
```
Yeah, that's a problem... :(

Random-Liu commented on pkg/server/sandbox_run.go:
```diff
@@ -241,10 +232,9 @@ func (c *criContainerdService) RunPodSandbox(ctx context.Context, r *runtime.Run
-	if err := c.sandboxStore.Create(meta); err != nil {
-		return nil, fmt.Errorf("failed to add sandbox metadata %+v into store: %v",
-			meta, err)
+	sandbox.CreatedAt = time.Now().UnixNano()
```
We don't maintain state of sandbox. We infer that directly from sandbox container state.

We could not do that for container because we also need exit code, finished timestamp etc. which we have to checkpoint ourselves. But for sandbox, we only care about whether it is running or not.

mikebrow replied:
ok yeah been having some debug issues on the update PR for sandboxes that have a messed up state (can't be stopped) then later on can't be created cause /pause isn't there.. the rootfs is messed up.

Random-Liu replied:
I'll help debug your PR over the weekend. :)

Random-Liu commented on pkg/server/container_list.go:
```diff
@@ -36,33 +34,31 @@ func (c *criContainerdService) ListContainers(ctx context.Context, r *runtime.Li
 	var containers []*runtime.Container
-	for _, meta := range metas {
-		containers = append(containers, toCRIContainer(meta))
+	for _, container := range containersInStore {
```
What do you mean here? :)

mikebrow replied:
you renamed range metas to range containersInStore presumably cause you didn't like the term meta... and didn't have a better option than just calling the data container again... :-)

Random-Liu commented on pkg/store/container/container.go:
```diff
@@ -0,0 +1,108 @@
+	// TODO(random-liu): Add stop channel to get rid of stop poll waiting.
+}
+
+// NewContainer creates a internal used container type.
```
Done.

Random-Liu commented on pkg/store/container/container.go:
```diff
@@ -0,0 +1,108 @@
+	}, nil
+}
+
+// LoadContainer loads the internal used container type.
```
Done.

Random-Liu commented:
@mikebrow Addressed comments.

The code could be cleaned up more, especially when we start to use containerd client. We could do that in the future.
